umich_api/api_utils.py

```diff
@@ -114,21 +122,44 @@ def _api_call(self, api_call, access_token, method="GET", payload=None):
             return resp
 
     def get_next_page(self, response):
+        #type: requests.Response -> Dict
         if not 'next' in response.links:
             return None
         self.__log.debug(response.links)
         return parse_qs(urlparse(response.links['next']['url']).query)
 
     def get_access_token(self, token_url, client_scope):
-        # type: (str) -> str
-        """Retrieves an access token from the specified URL
-        
+        # type: (str, str) -> str
+        """Retrieves access token from the local tokens dictionary
+           If it's expired it will attempt to renew it
         :param token_url: Full URL to retrieve access token
         :type token_url: str
         :param client_scope: Client scope to retrieve this token for
         :type client_scope: str
+        :returns Access token str
         """
+        
+        return self.get_oauth_token(token_url, client_scope).get("access_token")
 
+    def get_oauth_token(self, token_url, client_scope):
+        # type: (str, str) -> Dict
+        """Retrieves an access token from the specified URL, also caches the token in self.tokens dict
+        :param token_url: Full URL to retrieve access token
+        :type token_url: str
+        :param client_scope: Client scope to retrieve this token for
+        :type client_scope: str
+        :returns Dict (from json) generated from the response containing token information
+        """
+        if client_scope in self.tokens:
+            cached_token = self.tokens.get(client_scope)
+            # Check expires_time, if this is valid just return the token other-wise renew it
+            if (datetime.now() < cached_token.get("expires_at")):
```

Comment:
I went through your logic, I am confused about something. The way you are testing the expired token is by having the `expires_at` in the token dict this is fine for testing. will the `expires_at` sent anywhere as part of the API response? I did not see the `expires_at` parameter sent as part of the token response.  if a token is generated and you find in the list you use, but logic is checking for `expires_at` for testing purpose. Where is the case handled when token actually expired?

Reply:
Yeah this is an error and a great find! 

expires_in is actually the value that comes back from the token. 

I wanted to add a new calculated value (line 178) with expires_time for the time that this expires. However I messed up all of these values! :( 

I'll fix this and make a better test. Thanks for finding it! 🎊